sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/LogicalPlanStats.scala

```diff
@@ -18,33 +18,38 @@
 package org.apache.spark.sql.catalyst.plans.logical.statsEstimation
 
 import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.trees.TreeNodeTag
 
 /**
  * A trait to add statistics propagation to [[LogicalPlan]].
  */
 trait LogicalPlanStats { self: LogicalPlan =>
+  import LogicalPlanStats.STATS_CACHE_TAG
 
   /**
    * Returns the estimated statistics for the current logical plan node. Under the hood, this
    * method caches the return value, which is computed based on the configuration passed in the
    * first time. If the configuration changes, the cache can be invalidated by calling
    * [[invalidateStatsCache()]].
    */
-  def stats: Statistics = statsCache.getOrElse {
+  def stats: Statistics = statsOpt.getOrElse {
     if (conf.cboEnabled) {
-      statsCache = Option(BasicStatsPlanVisitor.visit(self))
+      setTagValue(STATS_CACHE_TAG, BasicStatsPlanVisitor.visit(self))
     } else {
-      statsCache = Option(SizeInBytesOnlyStatsPlanVisitor.visit(self))
+      setTagValue(STATS_CACHE_TAG, SizeInBytesOnlyStatsPlanVisitor.visit(self))
     }
-    statsCache.get
+    statsOpt.get
   }
 
-  /** A cache for the estimated statistics, such that it will only be computed once. */
-  protected var statsCache: Option[Statistics] = None
```

Comment:
it's fragile to use member variable to keep stats, as they will be lost after copy.